TombRaiderIV/Autosplitter.cs

```diff
@@ -1,91 +1,132 @@
-            return false;
-
-        // Handle all of TTE as well as TR4 ILs for both rulesets.
-        if (!Settings.FullGame || playingTheTimesExclusive)
+        // Handle IL / Section runs, which assumes all level transitions are desirable splits.
+        if (!Settings.FullGame)
         {
-            // This assumes all level transitions are desirable splits.
             bool loadingAnotherLevel = currentGfLevelComplete != 0;
             return loadingAnotherLevel;
         }
 
-        // Handle FG for each ruleset (TR4 only).
-        bool glitchless = Settings.Option;
-        return glitchless ? GlitchlessShouldSplit() : GlitchedShouldSplit();
+        // Handle Legacy Glitchless runs.
+        if (Settings.LegacyGlitchless)
+            return GlitchlessShouldSplit();
+
+        // Handle Full Game, non-Glitchless runs.
+        bool playingTheTimesExclusive = (Tr4Version)Data.GameVersion == Tr4Version.TheTimesExclusive;
+        return playingTheTimesExclusive ? TteShouldSplit() : Tr4ShouldSplit();
     }
 
-    private bool GlitchedShouldSplit()
+    /// <remarks>This assumes double-splits have already been prevented.</remarks>
+    private bool Tr4ShouldSplit()
     {
-        uint oldGfLevelComplete = Data.GfLevelComplete.Old;
-        bool leavingUnusedLevelValue = oldGfLevelComplete is 10 or 29;
-        if (leavingUnusedLevelValue)
+        uint nextLevel = Data.GfLevelComplete.Current;
+        uint currentLevel = Data.Level.Current;
+        if (nextLevel == currentLevel || nextLevel == 0)
             return false;
 
+        Tr4Level lower;
+        Tr4Level higher;
+        TransitionDirection direction;
+        bool currentRoomIsInLowerLevel;
+        if (nextLevel >= currentLevel)
+        {
+            direction = TransitionDirection.OneWayFromLower;
+            lower = (Tr4Level)currentLevel;
+            higher = (Tr4Level)nextLevel;
+            currentRoomIsInLowerLevel = true;
+        }
+        else
+        {
+            direction = TransitionDirection.OneWayFromHigher;
+            lower = (Tr4Level)nextLevel;
+            higher = (Tr4Level)currentLevel;
+            currentRoomIsInLowerLevel = false;
+        }
+
+        short room = Data.Room.Current;
+        var activeMatches = Settings
+            .Tr4LevelTransitions
+            .Where(t =>
+                t.Active &&
+                t.Lower == lower &&
+                t.Higher == higher &&
+                (t.Directionality == TransitionDirection.TwoWay || t.Directionality == direction) &&
+                t.RoomMatchedOrNotRequired(room, currentRoomIsInLowerLevel)
+            )
+            .ToList();
+
+        if (!activeMatches.Any())
+        {
+#if DEBUG
+            // Warn is the minimum threshold when using LiveSplit's Event Viewer logging.
+            LiveSplit.Options.Log.Warning($"No active transition match found!\nTransition: {lower} | {direction} | {higher} | room: {room}\n");
```

Comment:
Do LiveSplit logs actually use interpolated strings instead of format strings?

Reply:
LS's `Log` is a sort of wrapper over `System.Diagnostics.Trace`, which does use format strings:
```cs
public static class Log
{
    static Log()
    {
        try
        {
            if (!EventLog.SourceExists("LiveSplit"))
            {
                EventLog.CreateEventSource("LiveSplit", "Application");
            }
        }
        catch { }

        try
        {
            var listener = new EventLogTraceListener("LiveSplit")
            {
                Filter = new EventTypeFilter(SourceLevels.Warning)
            };
            Trace.Listeners.Add(listener);
        }
        catch { }
    }

    public static void Error(Exception ex)
    {
        try
        {
            Trace.TraceError("{0}\n\n{1}", ex.Message, ex.StackTrace);
        }
        catch { }
    }

    public static void Error(string message)
    {
        try
        {
            Trace.TraceError(message);
        }
        catch { }
    }

    <other methods>
}```

Reply:
Afaik, the usual definitions supporting structured/rich logging are like this:
```cs
Trace.TraceError(string format, params object[] args);
Trace.TraceInformation(string format, params object[] args);
Trace.TraceWarning(string format, params object[] args);
```

LS's wrapper does not provide any methods in this format. So I believe using structured logging while utilizing `LiveSplit.Options.Log` is not possible without some extension or re-implementation.

Per LS's implementation, I believe these `Trace` calls will only target Windows' EventViewer, as that describes their only usage of `Listeners.Add(...)`. In this context, I don't mind the lack of structured logging too much, so I don't believe related changes are necessary for us.